### PR TITLE
[Fix-5538]: Assign observer's current Result when optimistic read occurs

### LINKED ISSUE #5538: Why useEffect triggers twice after the API response is cached?
### Discussed in https://github.com/TanStack/query/discussions/5537

<div type='discussions-op-text'>

<sup>Originally posted by **dev-rish** June  5, 2023</sup>
**Describe the Bug**
After switching between 1 & 2 for couple of times (so that react query caches data for each of them), the useEffect still triggers twice.
The expected behavior is seen till v3.8.3 i.e. useEffect only triggers once. Once react query is updated to v3.9.0 every click on 1 or 2 causes useEffect to trigger twice.

Is there a way to prevent the second trigger?

I found related issues but still wanted some more info. if possible.

**Minimal, reproducible example**
v3: https://codesandbox.io/s/optimistic-firefly-1tk7i5
v4: https://codesandbox.io/s/sad-scott-cidk6s?file=/src/App.js

**Steps to reproduce**
Click in 1 > Click on 2 > Click in 1 > Click on 2 

**Expected behavior**
useEffect to trigger only once when `id` is switched.

**How often does this bug happen?**
Always</div>

**Note**:

- Seems to be `select` related, because it works as expected if `select` is removed
- This is _not_ strict mode related

### PR DESCRIPTION
**Pull request by incepter**
fixes #5538

### CODE REVIEW

File: packages/react-query-persist-client/src/__tests__/PersistQueryClientProvider.test.tsx [hunk 1]
```diff
@@ -340,9 +340,6 @@ describe('PersistQueryClientProvider', () => {
       fetchStatus: 'idle',
       data: 'hydrated',
     })
-
-    // #5443 seems like we get an extra render now ...
-    expect(states[1]).toStrictEqual(states[2])
```
Comment (TkDodo) [inline packages/react-query-persist-client/src/__tests__/PersistQueryClientProvider.test.tsx:345]: wow, this is great news 👏

File: packages/react-query/src/__tests__/useQuery.test.tsx [hunk 2]
```diff
@@ -5916,8 +5920,84 @@ describe('useQuery', () => {
...
+        setTimeout(
+          () =>
+            resolve({
+              numbers: {
+                current: {
+                  id,
+                },
+              },
+            }),
+          1000,
```
Comment (TkDodo) [inline packages/react-query/src/__tests__/useQuery.test.tsx:5943]: we usually try to have smaller timeouts in our tests please. Also, we can make this a bit shorter with our `sleep` util:

```
async function fetchNumber(id: number) {
  await sleep(50)
  return { numbers: { current: { id } } }
}
```

/edit: oh, I see you're using fake timers, so the amount of sleep probably doesn't matter. We don't have fake timers anywhere else I think, and in v5 (alpha branch), we've switched to `vitest`, so maybe we could just do a real sleep but for the shorter time?

**Review by TkDodo — COMMENTED** [review]
thank you so much 🙌

File: packages/react-query/src/__tests__/useQuery.test.tsx (hunk 2 above)
Comment (incepter) [reply packages/react-query/src/__tests__/useQuery.test.tsx:5943]: Yeah sure, I will remove the fake timers since they aren't use anywhere else (they work the same in vitest).

The real timeout leads some tests to have a random behavior depending on the local configuration, I encountered some local failing tests where setting a higher sleep value resolves it.

File: packages/query-core/src/queryObserver.ts [hunk 3]
```diff
@@ -240,7 +240,28 @@ export class QueryObserver<
...
+      // result, while we leave the `observer.currentResult`, so when new data
+      // arrives, it finds the old `observer.currentResult` which is related
+      // to the old QueryKey. Which means that currentResult and selectData are
+      // out of sync already.
+      // To solve this, we move the cursor of the currentResult everytime
+      // an observer reads an optimistic value.
+
+      // When keeping the previous data, the result doesn't change until new
+      // data arrives.
+      this.currentResult = result
```
[the remark this answers is not in the document]
Comment (incepter) [reply packages/query-core/src/queryObserver.ts:262]: These points are on spot, thanks a lot. (I knew I am ignoring lots of properties)

I will do a second pass later on my PR and resolve these points.
Can you please point the link towards the v5 changes ? Because I did not grasp the placeholder data behavior, is it only shown when pending and no cached data or always ? In what way placeholderData (or keepPrevious) affects the currentResult ?

[the remark this answers is not in the document]
Comment (incepter) [reply packages/query-core/src/queryObserver.ts:262]: I changed a little how code was presented and added the missing properties.

The function `shouldAssignObserverCurrentProperties` will be responsible for treating conditions for v4 and v5, I kept the independent `if elses` so removing the one related to `keepPreviousData` wouldn't do harm and would be straightforward.

**Comment by TkDodo** [comment]
I think we can ship this, but could you please also make the same PR to the `alpha` branch? I think this would be better than to merge main -> alpha and try to address the conflicts there

**Review by TkDodo — APPROVED** [review]

**Comment by incepter** [comment]
> I think we can ship this, but could you please also make the same PR to the `alpha` branch? I think this would be better than to merge main -> alpha and try to address the conflicts there

Yes sure, I ported the changes to alpha branch in this [PR](https://github.com/TanStack/query/pull/5611).

File: packages/query-core/src/queryObserver.ts [hunk 4]
```diff
@@ -764,3 +787,51 @@ function isStale(
...
+
+  // this means we want to put some placeholder data when pending and queryKey
+  // changed.
+  if (options.placeholderData !== undefined) {
+    // re-assign properties only if current data is placeholder data
+    // which means that data did not arrive yet, so, if there is some cached data
+    // we need to "prepare" to receive it
+    return optimisticResult.isPlaceholderData
+  }
+
```
[the remark this answers is not in the document]
Comment (TkDodo) [reply packages/query-core/src/queryObserver.ts:834]: I think it's fine like this. maybe a test where `placeholderData` is used would be good so that we're actually testing this branch?

File: packages/query-core/src/queryObserver.ts [hunk 5]
```diff
@@ -764,3 +787,51 @@ function isStale(
...
+    TQueryData,
+    TQueryKey
+  >,
+) {
+  // it is important to keep this condition like this for three reasons:
+  // 1. It will get removed in the v5
+  // 2. it reads: don't update the properties if we want to keep the previous
+  // data.
+  // 3. The opposite condition (!options.keepPreviousData) would fallthrough
+  // and will result in a bad decision
```
[the remark this answers is not in the document]
Comment (TkDodo) [reply packages/query-core/src/queryObserver.ts:815]: yeah I think you are right, especially when we consider that `createResult` is called during render (via `getOptimisticResult`), so we are actually having an observable side-effect during render.

We're now trying to "counteract" that side-effect by adding more writes during render 😅. Saving `selectResult` and `selectError` was done as a perf optimization, so that select wouldn't needlessly be invoked:

https://github.com/TanStack/query/blob/457723f2cd47a460d351d695d94bff962e7f68e5/packages/query-core/src/queryObserver.ts#L450-L456

For that, it's like priming a cache, so I think it should be fine from react's perspective. But if you have better ideas, I'm all ears. We do have quite good test coverage for that area, because it took some iterations to not get it wrong in various edge cases (and we added a test for each one of them).

File: packages/query-core/src/queryObserver.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (incepter) [reply packages/query-core/src/queryObserver.ts]: I added a test and also flipped the order of if elses to match the v5 ( or else, i was not able to reach that branch in the first place)

**Comment by TkDodo** [comment]
there is one test failing now @incepter:

> useQuery › should reuse same data object reference when queryKey changes and placeholderData is present

**Comment by incepter** [comment]
> there is one test failing now @incepter:
> 

Yes indeed @TkDodo , I fixed it. (I was using the same queryKey and same client for both tests, which resulted in them passing individually, then failing in ci.)
